code/hud/hudtarget.cpp

```diff
@@ -3979,46 +3970,117 @@ void HudGaugeLeadIndicator::renderLeadCurrentTarget()
+			auto bank = i;
+			// calculate the range of the weapon, and only display the lead target indicator
+			// if the weapon can actually hit the target
+			Assert(bank >= 0 && bank < swp->num_primary_banks);
+			Assert(swp->primary_bank_weapons[bank] < weapon_info_size());
+
+			if (swp->primary_bank_weapons[bank] < 0)
+				continue;
+
+			wip = &Weapon_info[swp->primary_bank_weapons[bank]];
+
+			if (wip->wi_flags[Weapon::Info_Flags::Nolink] ||  (wip->wi_flags[Weapon::Info_Flags::Ballistic] && swp->primary_bank_ammo[bank] <= 0))
+				continue;
+
+			if (bank < 0)
+				continue;
+
+			if (pm->n_guns && bank != -1 ) {
+				rel_pos = &pm->gun_banks[bank].pnt[0];
+			} else {
+				rel_pos = NULL;
+			}
+
+			source_pos = Player_obj->pos;
+			if (rel_pos != NULL) {
+				vec3d	gun_point;
+				vm_vec_unrotate(&gun_point, rel_pos, &Player_obj->orient);
+				vm_vec_add2(&source_pos, &gun_point);
+			}
+
+			auto weapon_range = MIN((wip->max_speed * wip->lifetime), wip->weapon_range);
+			if (weapon_range < dist_to_target)
+				continue;
+
+			frame_offset = pickFrame(weapon_range, srange, dist_to_target);
+
+			if (frame_offset >= 0) {
+				hud_calculate_lead_pos(&lead_target_pos, &target_pos, targetp, wip, dist_to_target);
+				if (Lead_indicator_behavior == leadIndicatorBehavior::MULTIPLE) {
+					renderIndicator(frame_offset, targetp, &lead_target_pos);
+				}
+				else if (Lead_indicator_behavior == leadIndicatorBehavior::AVERAGE) {
+					vm_vec_add2(&averaged_lead_pos, &lead_target_pos);
+					average_instances++;
+				}
+			}
+		}
+		if (Lead_indicator_behavior == leadIndicatorBehavior::AVERAGE) {
+			averaged_lead_pos = averaged_lead_pos/i2fl(average_instances);
+			renderIndicator(frame_offset, targetp, &averaged_lead_pos);
 		}
 	}
+	else {
+		if (pm->n_guns && bank_to_fire != -1 ) {
+			rel_pos = &pm->gun_banks[bank_to_fire].pnt[0];
+		} else {
+			rel_pos = NULL;
+		}
 
-	frame_offset = pickFrame(prange, srange, dist_to_target);
-	if ( frame_offset < 0 && srange != -1.0f ) {
-		return;
-	}
+		// source_pos will contain the world coordinate of where to base the lead indicator prediction
+		// from.  Normally, this will be the world pos of the gun turret of the currently selected primary
+		// weapon.
+		source_pos = Player_obj->pos;
+		if (rel_pos != NULL) {
+			vec3d	gun_point;
+			vm_vec_unrotate(&gun_point, rel_pos, &Player_obj->orient);
+			vm_vec_add2(&source_pos, &gun_point);
+		}
 
-	if (frame_offset >= 0) {
-		hud_calculate_lead_pos(&lead_target_pos, &target_pos, targetp, wip, dist_to_target);
-		renderIndicator(frame_offset, targetp, &lead_target_pos);
+		frame_offset = pickFrame(prange, srange, dist_to_target);
+		if ( frame_offset < 0 && srange != -1.0f ) {
+			return;
+		}
+
+		if (frame_offset >= 0) {
+			hud_calculate_lead_pos(&lead_target_pos, &target_pos, targetp, wip, dist_to_target);
+			renderIndicator(frame_offset, targetp, &lead_target_pos);
+		}
 	}
 
+	if ( bank_to_fire < 0 )
+		return;
+
```

Comment:
Since you moved this check down, can it happen that, in the default mode, an indicator is drawn for bank_to_fire == -1 before this check exists the function here?

Reply:
I think you're right, but this also makes me realize that the original check (as well as this one) also kills the dumbfire secondary lead indicator when `bank_to_fire == -1`. A player without primaries but with dumbfire secondaries would not see an indicator, it seems. Will fix.

Reply:
Yes. This _could_ be changed, but would have to be enabled with a game settings table, as it's a rather significant difference to retail

Reply:
Oh, that's fair

Reply:
One note that I realized afterwards by reading old code: 
Retail apparently never had dumbfire lead indicators. It was the first post-volition change to this file :laughing: 

But since it's so old, I think it still stands that this is the way to go